src/psyclone/psyir/transformations/acc_update_trans.py

```diff
@@ -0,0 +1,332 @@
+                    # Conservatively add an update host statement just before
+                    # the Call node for fear of temporary operands.
+                    self._add_update_directives([child])
+                elif isinstance(child, IfBlock):
+                    # Add any update statements that are required due to
+                    # (read) accesses within the condition of the If.
+                    self._add_update_directives([child.condition])
+                    # Recurse down into the if body
+                    self._add_updates_to_schedule(child.if_body)
+                    if child.else_body:
+                        self._add_updates_to_schedule(child.else_body)
+                elif isinstance(child, Loop):
+                    # A loop on the CPU with code that may be on the GPU.
+                    # The loop start, stop and step values could all
+                    # potentially have been previously written to on the GPU.
+                    self._add_update_directives([child.start_expr,
+                                                 child.stop_expr,
+                                                 child.step_expr])
+                    # Recurse down into the loop body
+                    self._add_updates_to_schedule(child.loop_body)
+
+        # We've reached the end of the list of children - are there any
+        # last nodes that represent computation on the CPU?
+        self._add_update_directives(node_list)
+
+    def _add_update_directives(self, node_list):
+        '''
+        Adds the required update directives before and after the nodes in
+        the supplied list.
+
+        :param sched: the schedule which contains the nodes in node_list.
+        :type sched: :py:class:`psyclone.psyir.nodes.Schedule`
+        :param node_list: the PSyIR nodes representing code executed on the \
+                          CPU rather than the GPU.
+        :type node_list: list of :py:class:`psyclone.psyir.nodes.Node`
+
+        '''
+        if not node_list:
+            return
+
+        inputs, outputs = self._dep_tools.get_in_out_parameters(node_list)
+        inputs, outputs = set(inputs), set(outputs)
+
+        # Workaround for lack of precise access descriptions, e.g. in the
+        # presence of array slices we currently overapproximate dependencies.
+        inputs.update(outputs)
+
+        # Workaround for CodeBlocks
+        # TODO We are probably relying on undefined behaviour here.
+        for node in node_list:
+            if isinstance(node, CodeBlock):
+                for symbol_name in node.get_symbol_names():
+                    inputs.add(Signature(symbol_name))
+                    outputs.add(Signature(symbol_name))
+
+        # Copy any data that is read by this region to the host (resp. device)
+        # if it is on the device (resp. host).
+        IN, OUT = 0, 1
+        for idx, cpu_sig in enumerate((inputs, outputs)):
```

Comment:
This is a complex and deeply-nested looping structure which, if I understand it correctly, mainly exists to avoid code duplication for the 'IN' and 'OUT' cases. I suspect it can be made clearer by refactoring and adding some utility routines that are just called with appropriate arguments instead.

Reply:
Hopefully the new comments and documentation should clarify the algorithm and alleviate the need for this. If that's not the case I'll have a second look.